mongosql/src/algebrizer/definitions.rs

```diff
@@ -475,14 +478,105 @@ impl<'a> Algebrizer<'a> {
         }
 
         // Build the Project Stage using the source and built expression.
-        let stage = mir::Stage::Project(mir::Project {
+        let project_stage = mir::Stage::Project(mir::Project {
             source: Box::new(source),
-            expression,
+            expression: expression.clone(),
             is_add_fields,
             cache: SchemaCache::new(),
         });
-        stage.schema(&self.schema_inference_state())?;
-        Ok(stage)
+        project_stage.schema(&self.schema_inference_state())?;
+        Ok(if is_distinct {
+            let group_stage = Self::build_distinct_values_group_and_project_stage(
+                project_stage,
+                self.scope_level,
+                datasources,
+            );
+            group_stage.schema(&self.schema_inference_state())?;
+            group_stage
+        } else {
+            project_stage
+        })
+    }
+
+    pub fn build_distinct_values_group_and_project_stage(
+        source_stage: mir::Stage,
+        scope_level: u16,
+        datasources: BTreeSet<Key>,
+    ) -> mir::Stage {
+        let mut group_keys = Vec::new();
+        let mut bottom: Option<AliasedExpr> = None;
+        let mut datasource_aliases = Vec::new();
+
+        // Identify data sources and build group keys for deduplication
+        for key in datasources.iter() {
+            if key.scope == scope_level {
+                match &key.datasource {
+                    DatasourceName::Named(name) => {
+                        group_keys.push(OptionallyAliasedExpr::Aliased(AliasedExpr {
+                            alias: name.clone(),
+                            expr: Expression::Reference(ReferenceExpr { key: key.clone() }),
+                        }));
+                        datasource_aliases.push(name.clone());
+                    }
+                    DatasourceName::Bottom => {
+                        bottom = Some(AliasedExpr {
+                            alias: BOTTOM_ALIAS.to_string(),
+                            expr: Expression::Reference(ReferenceExpr { key: key.clone() }),
+                        });
```

Comment:
Similar to what I said for the `*` case, we should (1) not special case `Bottom` and (2) use some other pattern for creating group key names since we're free to use whatever we want as long as we avoid naming collisions -- again, using datasource names _could_ result in a collision with `Bottom` in rare circumstances.

Reply:
Okay, I added a similar logic here as with the `*` case.